### PR TITLE
GH-40142: [Python] Allow FileInfo instances to be passed to dataset init

### LINKED ISSUE #40142: [Python] Allow Dataset construction from `FileInfo` objects
### Describe the enhancement requested

I'm developing a new dask integration with pyarrow parquet reader (see https://github.com/dask-contrib/dask-expr/pull/882) and want to rely on the pyarrow Filesystem more.

Right now, we are performing a list operation ourselves to get all touched files and I would like to pass the retrieved `FileInfo` objects directly to the dataset constructor. This API is already exposed in C++ but python bindings are missing.

The benefit of this is that there is API is that it cuts the need to perform additional HEAD requests to a remote storage.

This came up in https://github.com/apache/arrow/issues/38389#issuecomment-1774777681 and there's been related work already with https://github.com/apache/arrow/issues/37857

### Component(s)

Python

### PR DESCRIPTION
**Pull request by fjetter**
### Rationale for this change

Closes https://github.com/apache/arrow/issues/40142

I'm developing a new dask integration with pyarrow parquet reader (see https://github.com/dask-contrib/dask-expr/pull/882) and want to rely on the pyarrow Filesystem more.

Right now, we are performing a list operation ourselves to get all touched files and I would like to pass the retrieved `FileInfo` objects directly to the dataset constructor. This API is already exposed in C++ and this PR is adding the necessary python bindings.

The benefit of this is that there is API is that it cuts the need to perform additional HEAD requests to a remote storage.

This came up in https://github.com/apache/arrow/issues/38389#issuecomment-1774777681 and there's been related work already with https://github.com/apache/arrow/issues/37857

### What changes are included in this PR?

Python bindings for the `DatasetFactory` constructor that accepts a list/vector of `FileInfo` objects.

### Are these changes tested?

~I slightly modified the minio test setup such that the prometheus endpoint is exposed. This can be used to assert that there hasn't been any HEAD requests.~ I ended up removing this again since parsing the response is a bit brittle.

### Are there any user-facing changes?

* Closes: #40142

### CODE REVIEW

**Comment by fjetter** [comment]
the appveyor build seems to have timed out while compiling

**Comment by fjetter** [comment]
@jorisvandenbossche can I ask you to take a look at this?@jorisvandenbossche can I ask you to take a look at this?

File: python/pyarrow/tests/test_dataset.py [hunk 1]
```diff
@@ -2725,6 +2725,16 @@ def test_open_dataset_from_uri_s3(s3_example_simple, dataset_reader):
     assert dataset_reader.to_table(dataset).equals(table)
 
 
+@pytest.mark.parquet
+@pytest.mark.s3
+def test_dataset_from_fileinfos(s3_example_simple, dataset_reader):
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_dataset.py:2730]:
```suggestion
def test_open_dataset_from_fileinfos(s3_example_simple, dataset_reader):
```

(just to keep the naming consistent with the surrounding functions)

File: python/pyarrow/tests/conftest.py [hunk 2]
```diff
@@ -221,7 +221,7 @@ def minio_server_health_check(address):
         yield {
             'connection': s3_connection,
             'process': proc,
-            'tempdir': tmpdir
+            'tempdir': tmpdir,
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/conftest.py:224]:
```suggestion
            'tempdir': tmpdir
```

**Review by jorisvandenbossche — APPROVED** [review]
Thanks for the PR, sounds as a useful addition, and looks good!

**Comment by fjetter** [comment]
Thank you!

**Pull request merged**